packages/client/hmi-client/src/utils/pivot.ts

```diff
@@ -97,14 +99,84 @@ export const createMatrix1D = (data: any[]) => {
 			col: 0,
 			rowCriteria: data[rowIdx],
 			colCriteria: null,
-			value: data[rowIdx]
+			content: data[rowIdx]
 		});
 		rows.push(row);
 	}
 
 	return { matrix: rows };
 };
 
+export const createParameterMatrix = (
+	amr: Model,
+	transitionMatrixData: any[],
+	childParameterIds: string[]
+) => {
+	const rows: any[] = [];
```

Comment:
I think we can improve the construction logic to be more efficient 
- remove the input/output from the transitionMatrixData array, this makes the algo. more general
- resolve and cache transition's input/output
- resolve and cache input/output indices

Then we just need to loop over the transitionMatrixData twice, once to build the cache, and one more to populate the matrix, as oppose to looping for as many times as we have number of matrix-cells. 

The rough pseudo code is below:

```
createParameterMatrix = ( amr, transitionMatrixData, childParameterIds) => {
  const inputMap = Map<string, string>();
  const outputMap = Map<string, string>();

  // 1. Do the controller/input/output stuff
  for (let i = 0; i < transitionMatrixData.length; i++) {
    const transition = findTransition(amr, transitionMatrixData[i].id);

    // input/output/controller stuff

    inputMap.set(transitionMatrixData[i].id, ...);
    outputMap.set(transitionMatrixData[i].id, ...);
  }

  // 2. Build grid
  const rows: any[] = [];
  for (let rowIdx = 0; rowIdx < inputs.length; rowIdx++) {
    const row: PivotMatrixCell[] = [];
    for (let colIdx = 0; colIdx < outputs.length; colIdx++) {
      row.push({
        ...
      });
    }
  }

  const rowIndexMap = new Map();
  for (let rowIdx = 0; rowIdx < inputs.length; rowIdx++) { rowIndexMap.set(inputs[i], rowIdx); }

  const colIndexMap = new Map();
  for (let rowIdx = 0; rowIdx < inputs.length; rowIdx++) { colIndexMap.set(inputs[i], rowIdx); }

  // 3. Fill in the grid
  for (let i = 0; i < transitionMatrixData.length; i++) {
    const t = transitionMatrixData[i];
    const tInput = inputMap.get(t.id);
    const tOUtput = outputMap.get(t.id);

    const rowIdx = rowIndexMap.get(tInput);
    const colIdx = rowIndexMap.get(tOutput);
    rows[rowIdx][colIdx].content = ....
  }
}
```